File: 1898.maximum-number-of-removable-characters.py
#
# @lc app=leetcode id=1898 lang=python3
#
# [1898] Maximum Number of Removable Characters
#

# @lc code=start
from typing import List


class Solution:
    '''
    Then binary search
    Use isSubsequence
    '''
    def isSubsequence(self, mid) -> bool:
        p1, p2 = 0, 0
        while p1 < len(self.p) and p2 < len(self.s):
            if self.p[p1] == self.s[p2] and (p2 not in self.reverseRemovable or self.reverseRemovable[p2] >= mid):
                p1 += 1
            p2 += 1
        return p1 == len(self.p)
    

    def maximumRemovals(self, s: str, p: str, removable: List[int]) -> int:
        l, r = 0, len(removable)
        self.s = s
        self.p = p
        self.reverseRemovable = {}
        for i in range(len(removable)):
            self.reverseRemovable[removable[i]] = i
        while l < r:
            mid = l + (r - l) // 2
            if self.isSubsequence(mid):
                l = mid + 1
            else:
                r = mid
        if l == len(removable) and self.isSubsequence(l):
            return l
        else:
            return l - 1

# @lc code=end

File: 1898.maximum-number-of-removable-characters.cpp
/*
 * @lc app=leetcode id=1898 lang=cpp
 *
 * [1898] Maximum Number of Removable Characters
 */

// @lc code=start
#include <string>
#include <vector>
#include <unordered_map>

class Solution {
private:
    std::unordered_map<int, int> reverseRemovable;
public:
    bool isSubsequence(std::string s, std::string t, int mid) {
        int p1 = 0, p2 = 0;
        while (p1 < s.length() && p2 < t.length()) {
            if (s[p1] == t[p2] && (reverseRemovable.find(p2) == reverseRemovable.end() || reverseRemovable[p2] >= mid)) {
                p1++;
            }
            p2++;
        }
        return p1 == s.length();
    }

    int maximumRemovals(std::string s, std::string p, std::vector<int>& removable) {
        int l = 0, r = removable.size();
        reverseRemovable.clear();
        for (int i = 0; i < removable.size(); i++) {
            reverseRemovable[removable[i]] = i;
        }
        while (l < r) {
            int mid = l + (r - l) / 2;
            if (isSubsequence(p, s, mid)) {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        if (l == removable.size() && isSubsequence(p, s, l)) {
            return l;
        } else {
            return l - 1;
        }
    }
};

// @lc code=end

